gwin/calibration.py

```diff
@@ -0,0 +1,180 @@
+            the calibration options.
+        Return
+        ------
+        instance
+            An instance of the class.
+        """
+        all_params = dict(cp.items(section))
+        params = {key[len(ifo)+1:]: all_params[key]
+                  for key in all_params if ifo.lower() in key}
+        model = params.pop('model')
+        params['ifo_name'] = ifo.lower()
+
+        return all_models[model](**params)
+
+
+class CubicSpline(Recalibrate):
+
+    # spline model from https://dcc.ligo.org/LIGO-T1400682/public
+    name = 'cubic_spline'
+
+    def __init__(self, minimum_frequency, maximum_frequency, n_points,
+                 ifo_name):
+        """
+        Cubic spline recalibration
+
+        see https://dcc.ligo.org/LIGO-T1400682/public
+
+        This assumes the spline points follow
+        np.logspace(np.log(minimum_frequency), np.log(maximum_frequency),
+                    n_points)
+
+        Parameters
+        ----------
+        minimum_frequency: float
+            minimum frequency of spline points
+        maximum_frequency: float
+            maximum frequency of spline points
+        n_points: int
+            number of spline points
+        """
+        Recalibrate.__init__(self, ifo_name=ifo_name)
+        minimum_frequency = float(minimum_frequency)
+        maximum_frequency = float(maximum_frequency)
+        n_points = int(n_points)
+        if n_points < 4:
+            raise ValueError(
+                'Use at least 4 spline points for calibration model')
+        self.n_points = n_points
+        self.spline_points = np.logspace(np.log(minimum_frequency),
+                                         np.log(maximum_frequency), n_points)
+
+    def apply_calibration(self, strain):
+        """Apply calibration model
+
+        This applies cubic spline calibration to the strain.
+
+        Parameters
+        ----------
+        strain : FrequencySeries
+            The strain to be recalibrated.
+
+        Return
+        ------
+        strain_adjusted : FrequencySeries
+            The recalibrated strain.
+        """
+        amplitude_parameters =\
+            [self.params['amplitude_{}_{}'.format(self.ifo_name, ii)]
+             for ii in range(self.n_points)]
+        amplitude_spline = UnivariateSpline(self.spline_points,
+                                            amplitude_parameters)
+        delta_amplitude = amplitude_spline(strain.sample_frequencies.numpy())
+
+        phase_parameters =\
+            [self.params['phase_{}_{}'.format(self.ifo_name, ii)]
+             for ii in range(self.n_points)]
+        phase_spline = UnivariateSpline(self.spline_points, phase_parameters)
+        delta_phase = phase_spline(strain.sample_frequencies.numpy())
+
+        strain_adjusted = strain * (1.0 + delta_amplitude)\
+            * (2.0 + 1j * delta_phase) / (2.0 - 1j * delta_phase)
+
+        return strain_adjusted
+
+
+all_models = {
+    Recalibrate.name: Recalibrate,
```

Comment:
Don't include `Recalibrate` in `all_models`, since it can't actually be implemented.